File: ui/ios/Sources/SamView.swift
import SwiftUI
import SAMKit
import Combine

/// Interactive view for SAM segmentation on iOS
public struct SamView: View {
    @StateObject private var viewModel: SamViewModel
    @State private var currentImage: UIImage?
    @State private var scale: CGFloat = 1.0
    @State private var lastScale: CGFloat = 1.0
    @State private var offset: CGSize = .zero
    @State private var lastOffset: CGSize = .zero
    
    private let debounceDelay: Double = 0.08 // 80ms debounce
    
    public init(
        image: UIImage,
        model: SamModelRef,
        config: RuntimeConfig = .bestAvailable
    ) {
        self.currentImage = image
        self._viewModel = StateObject(wrappedValue: SamViewModel(
            image: image,
            model: model,
            config: config
        ))
    }
    
    public var body: some View {
        GeometryReader { geometry in
            ZStack {
                // Background
                Color.black.edgesIgnoringSafeArea(.all)
                
                // Image and mask overlay
                if let image = currentImage {
                    ImageCanvas(
                        image: image,
                        masks: viewModel.masks,
                        points: viewModel.points,
                        box: viewModel.currentBox,
                        selectedMaskIndex: viewModel.selectedMaskIndex,
                        scale: scale,
                        offset: offset,
                        onTap: handleTap,
                        onDrag: handleDrag
                    )
                }
                
                // Controls overlay
                VStack {
                    Spacer()
                    ControlPanel(viewModel: viewModel)
                        .padding()
                }
            }
            .gesture(magnificationGesture)
            .gesture(dragGesture)
        }
        .onAppear {
            viewModel.setImage(currentImage!)
        }
    }
    
    // MARK: - Gestures
    
    private var magnificationGesture: some Gesture {
        MagnificationGesture()
            .onChanged { value in
                scale = lastScale * value
            }
            .onEnded { _ in
                lastScale = scale
            }
    }
    
    private var dragGesture: some Gesture {
        DragGesture()
            .onChanged { value in
                offset = CGSize(
                    width: lastOffset.width + value.translation.width,
                    height: lastOffset.height + value.translation.height
                )
            }
            .onEnded { _ in
                lastOffset = offset
            }
    }
    
    private func handleTap(location: CGPoint) {
        let imagePoint = convertToImageCoordinates(location)
        viewModel.addPoint(at: imagePoint)
    }
    
    private func handleDrag(start: CGPoint, end: CGPoint) {
        let startImage = convertToImageCoordinates(start)
        let endImage = convertToImageCoordinates(end)
        
        viewModel.setBox(
            from: startImage,
            to: endImage
        )
    }
    
    private func convertToImageCoordinates(_ point: CGPoint) -> CGPoint {
        // Convert from view coordinates to image coordinates
        // accounting for scale and offset
        guard let image = currentImage else { return point }
        
        let imageSize = image.size
        let scaledSize = CGSize(
            width: imageSize.width * scale,
            height: imageSize.height * scale
        )
        
        let adjustedPoint = CGPoint(
            x: (point.x - offset.width) / scale,
            y: (point.y - offset.height) / scale
        )
        
        return adjustedPoint
    }
}

// MARK: - Image Canvas

struct ImageCanvas: View {
    let image: UIImage
    let masks: [SamMask]
    let points: [SamPoint]
    let box: SamBox?
    let selectedMaskIndex: Int
    let scale: CGFloat
    let offset: CGSize
    let onTap: (CGPoint) -> Void
    let onDrag: (CGPoint, CGPoint) -> Void
    
    @State private var dragStart: CGPoint?
    
    var body: some View {
        Canvas { context, size in
            // Draw image
            if let cgImage = image.cgImage {
                context.draw(
                    Image(cgImage, scale: 1.0, label: Text("Image")),
                    in: CGRect(origin: .zero, size: size)
                )
            }
            
            // Draw masks
            for (index, mask) in masks.enumerated() {
                if index == selectedMaskIndex || selectedMaskIndex == -1 {
                    drawMask(mask, in: context, size: size, opacity: 0.5)
                }
            }
            
            // Draw points
            for point in points {
                drawPoint(point, in: context, size: size)
            }
            
            // Draw box
            if let box = box {
                drawBox(box, in: context, size: size)
            }
        }
        .scaleEffect(scale)
        .offset(offset)
        .onTapGesture { location in
            onTap(location)
        }
        .gesture(
            DragGesture()
                .onChanged { value in
                    if dragStart == nil {
                        dragStart = value.startLocation
                    }
                }
                .onEnded { value in
                    if let start = dragStart {
                        onDrag(start, value.location)
                        dragStart = nil
                    }
                }
        )
    }
    
    private func drawMask(_ mask: SamMask, in context: GraphicsContext, size: CGSize, opacity: Double) {
        guard let maskImage = mask.toColoredCGImage(
            color: CGColor(red: 0, green: 0.5, blue: 1, alpha: 1)
        ) else { return }
        
        context.opacity = opacity
        context.draw(
            Image(maskImage, scale: 1.0, label: Text("Mask")),
            in: CGRect(origin: .zero, size: size)
        )
        context.opacity = 1.0
    }
    
    private func drawPoint(_ point: SamPoint, in context: GraphicsContext, size: CGSize) {
        let color = point.label == 1 ? Color.green : Color.red
        let position = CGPoint(x: CGFloat(point.x), y: CGFloat(point.y))
        
        context.fill(
            Circle().path(in: CGRect(
                x: position.x - 5,
                y: position.y - 5,
                width: 10,
                height: 10
            )),
            with: .color(color)
        )
    }
    
    private func drawBox(_ box: SamBox, in context: GraphicsContext, size: CGSize) {
        let rect = CGRect(
            x: CGFloat(box.x0),
            y: CGFloat(box.y0),
            width: CGFloat(box.x1 - box.x0),
            height: CGFloat(box.y1 - box.y0)
        )
        
        context.stroke(
            Rectangle().path(in: rect),
            with: .color(.yellow),
            lineWidth: 2
        )
    }
}

// MARK: - Control Panel

struct ControlPanel: View {
    @ObservedObject var viewModel: SamViewModel
    
    var body: some View {
        VStack(spacing: 12) {
            // Mask selection
            if viewModel.masks.count > 1 {
                Picker("Mask", selection: $viewModel.selectedMaskIndex) {
                    Text("All").tag(-1)
                    ForEach(0..<viewModel.masks.count, id: \.self) { index in
                        Text("Mask \(index + 1)").tag(index)
                    }
                }
                .pickerStyle(SegmentedPickerStyle())
            }
            
            // Threshold slider
            HStack {
                Text("Threshold")
                Slider(
                    value: $viewModel.maskThreshold,
                    in: -10...10,
                    onEditingChanged: { editing in
                        if !editing {
                            viewModel.updateThreshold()
                        }
                    }
                )
                Text(String(format: "%.1f", viewModel.maskThreshold))
            }
            
            // Action buttons
            HStack(spacing: 16) {
                Button("Clear Points") {
                    viewModel.clearPoints()
                }
                
                Button("Clear Box") {
                    viewModel.clearBox()
                }
                
                Button("Undo") {
                    viewModel.undo()
                }
                .disabled(!viewModel.canUndo)
                
                Button("Export") {
                    viewModel.exportMask()
                }
                .disabled(viewModel.masks.isEmpty)
            }
            .buttonStyle(.bordered)
        }
        .padding()
        .background(Color.black.opacity(0.8))
        .cornerRadius(12)
    }
}

// MARK: - View Model

@MainActor
class SamViewModel: ObservableObject {
    @Published var masks: [SamMask] = []
    @Published var points: [SamPoint] = []
    @Published var currentBox: SamBox?
    @Published var selectedMaskIndex: Int = -1
    @Published var maskThreshold: Float = 0.0
    @Published var isProcessing: Bool = false
    @Published var canUndo: Bool = false
    
    private var session: SamSession?
    private let model: SamModelRef
    private let config: RuntimeConfig
    private var currentImage: UIImage?
    private var debounceTimer: Timer?
    private var history: [(points: [SamPoint], box: SamBox?)] = []
    
    init(image: UIImage, model: SamModelRef, config: RuntimeConfig) {
        self.currentImage = image
        self.model = model
        self.config = config
        setupSession()
    }
    
    private func setupSession() {
        do {
            session = try SamSession(model: model, config: config)
        } catch {
            print("Failed to create session: \(error)")
        }
    }
    
    func setImage(_ image: UIImage) {
        currentImage = image
        clearAll()
        
        guard let cgImage = image.cgImage else { return }
        
        Task {
            isProcessing = true
            do {
                try session?.setImage(cgImage)
            } catch {
                print("Failed to set image: \(error)")
            }
            isProcessing = false
        }
    }
    
    func addPoint(at location: CGPoint) {
        let point = SamPoint(
            x: Float(location.x),
            y: Float(location.y),
            label: 1 // Default to positive
        )
        points.append(point)
        saveHistory()
        debouncedPredict()
    }
    
    func setBox(from start: CGPoint, to end: CGPoint) {
        currentBox = SamBox(
            x0: Float(min(start.x, end.x)),
            y0: Float(min(start.y, end.y)),
            x1: Float(max(start.x, end.x)),
            y1: Float(max(start.y, end.y))
        )
        saveHistory()
        debouncedPredict()
    }
    
    func clearPoints() {
        points.removeAll()
        debouncedPredict()
    }
    
    func clearBox() {
        currentBox = nil
        debouncedPredict()
    }
    
    func clearAll() {
        points.removeAll()
        currentBox = nil
        masks.removeAll()
        history.removeAll()
        canUndo = false
    }
    
    func undo() {
        guard history.count > 1 else { return }
        history.removeLast()
        if let last = history.last {
            points = last.points
            currentBox = last.box
            debouncedPredict()
        }
        canUndo = history.count > 1
    }
    
    func updateThreshold() {
        // Re-apply threshold to existing masks if we have logits
        // This would require keeping the original logits
        debouncedPredict()
    }
    
    func exportMask() {
        guard let mask = selectedMask else { return }
        // Export logic here - save to photos, share, etc.
        if let cgImage = mask.toCGImage() {
            UIImageWriteToSavedPhotosAlbum(
                UIImage(cgImage: cgImage),
                nil,
                nil,
                nil
            )
        }
    }
    
    private var selectedMask: SamMask? {
        if selectedMaskIndex == -1 && !masks.isEmpty {
            return masks.first
        } else if selectedMaskIndex >= 0 && selectedMaskIndex < masks.count {
            return masks[selectedMaskIndex]
        }
        return nil
    }
    
    private func saveHistory() {
        history.append((points: points, box: currentBox))
        canUndo = history.count > 1
    }
    
    private func debouncedPredict() {
        debounceTimer?.invalidate()
        debounceTimer = Timer.scheduledTimer(withTimeInterval: 0.08, repeats: false) { _ in
            Task { @MainActor in
                await self.predict()
            }
        }
    }
    
    private func predict() async {
        guard !points.isEmpty || currentBox != nil else {
            masks.removeAll()
            return
        }
        
        isProcessing = true
        
        do {
            let result = try await session?.predict(
                points: points,
                box: currentBox,
                maskInput: nil,
                options: SamOptions(
                    multimaskOutput: true,
                    returnLogits: true,
                    maskThreshold: maskThreshold
                )
            )
            
            if let result = result {
                masks = result.masks
            }
        } catch {
            print("Prediction failed: \(error)")
        }
        
        isProcessing = false
    }
}

File: ui/android/samkit-ui/src/main/java/com/samkit/ui/SamView.kt
package com.samkit.ui

import android.graphics.Bitmap
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.gestures.rememberTransformableState
import androidx.compose.foundation.gestures.transformable
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.*
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.samkit.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Interactive view for SAM segmentation on Android
 */
@Composable
fun SamView(
    bitmap: Bitmap,
    model: SamModelRef,
    config: RuntimeConfig = RuntimeConfig.Best,
    modifier: Modifier = Modifier
) {
    val viewModel: SamViewModel = viewModel {
        SamViewModel(bitmap, model, config)
    }
    
    val uiState by viewModel.uiState.collectAsState()
    
    var scale by remember { mutableStateOf(1f) }
    var offset by remember { mutableStateOf(Offset.Zero) }
    
    Box(modifier = modifier.fillMaxSize()) {
        // Image canvas with masks
        ImageCanvas(
            bitmap = bitmap,
            masks = uiState.masks,
            points = uiState.points,
            box = uiState.currentBox,
            selectedMaskIndex = uiState.selectedMaskIndex,
            scale = scale,
            offset = offset,
            onTap = { location ->
                viewModel.addPoint(location)
            },
            onDrag = { start, end ->
                viewModel.setBox(start, end)
            },
            onTransform = { scaleChange, offsetChange ->
                scale *= scaleChange
                offset += offsetChange
            },
            modifier = Modifier.fillMaxSize()
        )
        
        // Loading indicator
        if (uiState.isProcessing) {
            CircularProgressIndicator(
                modifier = Modifier
                    .align(Alignment.Center)
                    .size(48.dp)
            )
        }
        
        // Control panel
        ControlPanel(
            viewModel = viewModel,
            uiState = uiState,
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(16.dp)
        )
    }
    
    LaunchedEffect(bitmap) {
        viewModel.setImage(bitmap)
    }
}

@Composable
private fun ImageCanvas(
    bitmap: Bitmap,
    masks: List<SamMask>,
    points: List<SamPoint>,
    box: SamBox?,
    selectedMaskIndex: Int,
    scale: Float,
    offset: Offset,
    onTap: (Offset) -> Unit,
    onDrag: (Offset, Offset) -> Unit,
    onTransform: (Float, Offset) -> Unit,
    modifier: Modifier = Modifier
) {
    val density = LocalDensity.current
    var dragStart by remember { mutableStateOf<Offset?>(null) }
    
    val transformableState = rememberTransformableState { zoomChange, offsetChange, _ ->
        onTransform(zoomChange, offsetChange)
    }
    
    Canvas(
        modifier = modifier
            .transformable(state = transformableState)
            .pointerInput(Unit) {
                detectTapGestures { offset ->
                    val imagePoint = convertToImageCoordinates(
                        offset, scale, this@Canvas.offset, bitmap.width, bitmap.height
                    )
                    onTap(imagePoint)
                }
            }
            .pointerInput(Unit) {
                detectDragGestures(
                    onDragStart = { offset ->
                        dragStart = offset
                    },
                    onDragEnd = {
                        dragStart?.let { start ->
                            val end = start // Last drag position
                            val startImage = convertToImageCoordinates(
                                start, scale, this@Canvas.offset, bitmap.width, bitmap.height
                            )
                            val endImage = convertToImageCoordinates(
                                end, scale, this@Canvas.offset, bitmap.width, bitmap.height
                            )
                            onDrag(startImage, endImage)
                        }
                        dragStart = null
                    }
                ) { _, _ ->
                    // Track drag progress if needed
                }
            }
    ) {
        drawIntoCanvas { canvas ->
            // Apply transformations
            canvas.save()
            canvas.scale(scale)
            canvas.translate(offset.x, offset.y)
            
            // Draw bitmap
            val imageBitmap = bitmap.asImageBitmap()
            canvas.drawImage(
                image = imageBitmap,
                topLeftOffset = Offset.Zero
            )
            
            // Draw masks
            masks.forEachIndexed { index, mask ->
                if (index == selectedMaskIndex || selectedMaskIndex == -1) {
                    drawMask(mask, canvas, 0.5f)
                }
            }
            
            // Draw points
            points.forEach { point ->
                drawPoint(point, canvas)
            }
            
            // Draw box
            box?.let {
                drawBox(it, canvas)
            }
            
            canvas.restore()
        }
    }
}

private fun DrawScope.drawMask(mask: SamMask, canvas: androidx.compose.ui.graphics.Canvas, opacity: Float) {
    val maskBitmap = mask.toColoredBitmap(Color.Blue.toArgb()).asImageBitmap()
    val paint = Paint().apply {
        alpha = opacity
    }
    canvas.drawImage(
        image = maskBitmap,
        topLeftOffset = Offset.Zero,
        paint = paint
    )
}

private fun DrawScope.drawPoint(point: SamPoint, canvas: androidx.compose.ui.graphics.Canvas) {
    val color = if (point.label == 1) Color.Green else Color.Red
    val paint = Paint().apply {
        this.color = color
        style = PaintingStyle.Fill
    }
    
    canvas.drawCircle(
        center = Offset(point.x, point.y),
        radius = 5f,
        paint = paint
    )
}

private fun DrawScope.drawBox(box: SamBox, canvas: androidx.compose.ui.graphics.Canvas) {
    val paint = Paint().apply {
        color = Color.Yellow
        style = PaintingStyle.Stroke
        strokeWidth = 2f
    }
    
    canvas.drawRect(
        offset = Offset(box.x0, box.y0),
        size = Size(box.x1 - box.x0, box.y1 - box.y0),
        paint = paint
    )
}

private fun convertToImageCoordinates(
    viewPoint: Offset,
    scale: Float,
    offset: Offset,
    imageWidth: Int,
    imageHeight: Int
): Offset {
    return Offset(
        x = (viewPoint.x - offset.x) / scale,
        y = (viewPoint.y - offset.y) / scale
    )
}

@Composable
private fun ControlPanel(
    viewModel: SamViewModel,
    uiState: SamUiState,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.9f)
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            // Mask selection
            if (uiState.masks.size > 1) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceEvenly
                ) {
                    FilterChip(
                        selected = uiState.selectedMaskIndex == -1,
                        onClick = { viewModel.selectMask(-1) },
                        label = { Text("All") }
                    )
                    uiState.masks.forEachIndexed { index, _ ->
                        FilterChip(
                            selected = uiState.selectedMaskIndex == index,
                            onClick = { viewModel.selectMask(index) },
                            label = { Text("Mask ${index + 1}") }
                        )
                    }
                }
            }
            
            // Threshold slider
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text("Threshold", modifier = Modifier.width(80.dp))
                Slider(
                    value = uiState.maskThreshold,
                    onValueChange = { viewModel.updateThreshold(it) },
                    valueRange = -10f..10f,
                    modifier = Modifier.weight(1f)
                )
                Text(
                    String.format("%.1f", uiState.maskThreshold),
                    modifier = Modifier.width(40.dp)
                )
            }
            
            // Action buttons
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceEvenly
            ) {
                TextButton(onClick = { viewModel.clearPoints() }) {
                    Text("Clear Points")
                }
                TextButton(onClick = { viewModel.clearBox() }) {
                    Text("Clear Box")
                }
                TextButton(
                    onClick = { viewModel.undo() },
                    enabled = uiState.canUndo
                ) {
                    Text("Undo")
                }
                TextButton(
                    onClick = { viewModel.exportMask() },
                    enabled = uiState.masks.isNotEmpty()
                ) {
                    Text("Export")
                }
            }
        }
    }
}

// View Model
class SamViewModel(
    private val initialBitmap: Bitmap,
    private val model: SamModelRef,
    private val config: RuntimeConfig
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(SamUiState())
    val uiState: StateFlow<SamUiState> = _uiState.asStateFlow()
    
    private var session: SamSession? = null
    private var debounceJob: Job? = null
    private val history = mutableListOf<HistoryEntry>()
    
    init {
        initSession()
    }
    
    private fun initSession() {
        viewModelScope.launch {
            try {
                session = SamSession(model, config)
            } catch (e: Exception) {
                // Handle error
            }
        }
    }
    
    fun setImage(bitmap: Bitmap) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isProcessing = true)
            try {
                session?.setImage(bitmap)
            } catch (e: Exception) {
                // Handle error
            }
            _uiState.value = _uiState.value.copy(isProcessing = false)
        }
    }
    
    fun addPoint(location: Offset) {
        val point = SamPoint(location.x, location.y, 1)
        val newPoints = _uiState.value.points + point
        _uiState.value = _uiState.value.copy(points = newPoints)
        saveHistory()
        debouncedPredict()
    }
    
    fun setBox(start: Offset, end: Offset) {
        val box = SamBox(
            x0 = minOf(start.x, end.x),
            y0 = minOf(start.y, end.y),
            x1 = maxOf(start.x, end.x),
            y1 = maxOf(start.y, end.y)
        )
        _uiState.value = _uiState.value.copy(currentBox = box)
        saveHistory()
        debouncedPredict()
    }
    
    fun clearPoints() {
        _uiState.value = _uiState.value.copy(points = emptyList())
        debouncedPredict()
    }
    
    fun clearBox() {
        _uiState.value = _uiState.value.copy(currentBox = null)
        debouncedPredict()
    }
    
    fun selectMask(index: Int) {
        _uiState.value = _uiState.value.copy(selectedMaskIndex = index)
    }
    
    fun updateThreshold(threshold: Float) {
        _uiState.value = _uiState.value.copy(maskThreshold = threshold)
        debouncedPredict()
    }
    
    fun undo() {
        if (history.size > 1) {
            history.removeLast()
            history.lastOrNull()?.let { entry ->
                _uiState.value = _uiState.value.copy(
                    points = entry.points,
                    currentBox = entry.box
                )
                debouncedPredict()
            }
        }
        updateUndoState()
    }
    
    fun exportMask() {
        val masks = _uiState.value.masks
        val selectedIndex = _uiState.value.selectedMaskIndex
        val mask = if (selectedIndex >= 0 && selectedIndex < masks.size) {
            masks[selectedIndex]
        } else {
            masks.firstOrNull()
        }
        
        mask?.let {
            // Export logic - save to gallery, share, etc.
            val bitmap = it.toBitmap()
            // Save bitmap
        }
    }
    
    private fun saveHistory() {
        history.add(
            HistoryEntry(
                points = _uiState.value.points,
                box = _uiState.value.currentBox
            )
        )
        updateUndoState()
    }
    
    private fun updateUndoState() {
        _uiState.value = _uiState.value.copy(canUndo = history.size > 1)
    }
    
    private fun debouncedPredict() {
        debounceJob?.cancel()
        debounceJob = viewModelScope.launch {
            delay(80) // 80ms debounce
            predict()
        }
    }
    
    private suspend fun predict() {
        val points = _uiState.value.points
        val box = _uiState.value.currentBox
        
        if (points.isEmpty() && box == null) {
            _uiState.value = _uiState.value.copy(masks = emptyList())
            return
        }
        
        _uiState.value = _uiState.value.copy(isProcessing = true)
        
        try {
            val result = session?.predict(
                points = points,
                box = box,
                maskInput = null,
                options = SamOptions(
                    multimaskOutput = true,
                    returnLogits = true,
                    maskThreshold = _uiState.value.maskThreshold
                )
            )
            
            result?.let {
                _uiState.value = _uiState.value.copy(masks = it.masks)
            }
        } catch (e: Exception) {
            // Handle error
        }
        
        _uiState.value = _uiState.value.copy(isProcessing = false)
    }
    
    override fun onCleared() {
        super.onCleared()
        session?.close()
    }
}

data class SamUiState(
    val masks: List<SamMask> = emptyList(),
    val points: List<SamPoint> = emptyList(),
    val currentBox: SamBox? = null,
    val selectedMaskIndex: Int = -1,
    val maskThreshold: Float = 0f,
    val isProcessing: Boolean = false,
    val canUndo: Boolean = false
)

private data class HistoryEntry(
    val points: List<SamPoint>,
    val box: SamBox?
)